optimize/convex/lp/parametric.go

```diff
@@ -0,0 +1,448 @@
+package lp
+
+import (
+	"errors"
+	"golang.org/x/exp/rand"
+	"math"
+
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/mat"
+)
+
+var (
+	ErrDegenerate = errors.New("lp: current perturbation is degenerate")
+)
+
+const (
+	// absZeroTol is the absolute tolerance on testing if the elements of an
+	// update vector are zero.
+	absZeroTol = 1e-12
+	// relZeroTol is the tolerance on testing if the elements of an update
+	// vector are zero relative to the current values.
+	relZeroTol = 1e-12
+	// xZeroTol is the tolerance on testing if the elements of the
+	// optimal solution vector are zero (degeneracy).
+	xZeroTol = 1e-14
+	// swapCap is the default max capacity for the Swap object
+	swapCap = 25
+	// swapCondTol is the tolerence on the condition bound of the swap matrix;
+	// this is lower than mat.ConditionTolerance as to avoid the accumalation
+	// of "decimal dust" through repeated use of an ill-conditioned operator
+	swapCondTol = 1e8
+)
+
+// Trace stores the per-iteration output of the parametric simplex method.
+type Trace struct {
+	Lambda, X []float64
+	Idx       []int
+}
+
+// Parametric solves a linear program in standard form using the parametric
+// simplex method, a primal-dual algorithm. A description of the algorithm can
+// be found in Pang, H. et al. (2017). "Parametric Simplex Method for Sparse Learning".
+// Advances in Neural Information Processing Systems(30): p. 187-196.
+func Parametric(c []float64, A mat.Matrix, b []float64, tol float64, initialBasic []int, dense bool, rnd *rand.Rand) (optF float64, optX []float64, basicIdxs []int, err error) {
+	ans, x, idxs, _, err := parametric(initialBasic, c, A, b, nil, nil, tol, false, dense, rnd)
+	if dense {
+		return ans, x, nil, err
+	}
+	return ans, x, idxs, err
+}
+
+// ParametricWithTrace solves a linear program in standard form using the parametric
+// simplex method and returns the output of each iteration in a Trace object. This
+// function is useful for homotopy in L1 regularized minimization problems. Note that
+//  minimize	L(θ) + λ‖θ‖_1
+// can be written equivalently as
+//  minimize ‖θ‖_1
+//  s.t. 		‖∇L(θ)‖_∞ ≤ λ.
+// Either fomulation can both be expressed as a linear program when L(θ)
+// is piecewise linear or quadratic, respectively. Therefore the parametric simplex
+// method allows for recovery of the entire regularization path of at the cost of
+// solving a single linear program.
+func ParametricWithTrace(c []float64, A mat.Matrix, b, cbar, bbar []float64, tol float64, initialBasic []int, rnd *rand.Rand) (optTr *Trace, err error) {
```

Comment:
Is there a reason not to return the optimal solution as well?

Reply:
Mostly because that solution is the final entry in the `Path` do it seemed redundant. Not to mention when using the method for minimization, all solutions are optimal for different values of the regularization parameter.